From ad85b2c87e4f7b3b5e94d841e6d297814be9427b Mon Sep 17 00:00:00 2001
From: backslashxx <118538522+backslashxx@users.noreply.github.com>
Date: Tue, 3 Mar 2026 13:56:06 +0800
Subject: [PATCH] selinux: avc/selinuxfs: add selinux_hide hooks

Hook 'slow_avc_audit' to allow AVC audit filtering and 'selinux_transaction_write' to intercept SELinux context transactions through selinuxfs.
Note that 'slow_avc_audit' hook is optional and can be handled from userspace.
It is not part of standard KernelSU functionality but is included for implementations that require additional SELinux hiding support.

This implementation follows the minimal manual hook approach
documented by backslashxx and keeps CONFIG_KSU_LSM_SECURITY_HOOKS
enabled without requiring additional setprocattr hooks.

Reference:
- https://github.com/backslashxx/KernelSU/issues/5 (slow_avc_audit and selinux_transaction_write hooks)

Co-authored-by: backslashxx <118538522+backslashxx@users.noreply.github.com>
Change-Id: Idfc79dc6bb936d05d5ebb1379a5ed3eb194b672e
Signed-off-by: Neebe3289 <neebexd@gmail.com>
[Neebe3289: rewrite commit message for clarity]
---
 security/selinux/avc.c       | 7 +++++++
 security/selinux/selinuxfs.c | 8 ++++++++
 2 files changed, 15 insertions(+)

diff --git a/security/selinux/avc.c b/security/selinux/avc.c
index 4e9d4893e67c..3468ccaa1153 100644
--- a/security/selinux/avc.c
+++ b/security/selinux/avc.c
@@ -777,6 +777,10 @@ static void avc_audit_post_callback(struct audit_buffer *ab, void *a)
 	}
 }
 
+#if defined(CONFIG_KSU) && !defined(CONFIG_KPROBES)
+extern void ksu_slow_avc_audit(u32 *tsid);
+#endif
+
 /* This is the slow part of avc audit with big stack footprint */
 noinline int slow_avc_audit(struct selinux_state *state,
 			    u32 ssid, u32 tsid, u16 tclass,
@@ -787,6 +791,9 @@ noinline int slow_avc_audit(struct selinux_state *state,
 	struct common_audit_data stack_data;
 	struct selinux_audit_data sad;
 
+#if defined(CONFIG_KSU) && !defined(CONFIG_KPROBES)
+	ksu_slow_avc_audit(&tsid);
+#endif
 	if (!a) {
 		a = &stack_data;
 		a->type = LSM_AUDIT_DATA_NONE;
diff --git a/security/selinux/selinuxfs.c b/security/selinux/selinuxfs.c
index 816b515e53ad..c698ae1c57aa 100644
--- a/security/selinux/selinuxfs.c
+++ b/security/selinux/selinuxfs.c
@@ -776,6 +776,10 @@ static ssize_t (*write_op[])(struct file *, char *, size_t) = {
 	[SEL_CONTEXT] = sel_write_context,
 };
 
+#if defined(CONFIG_KSU) && !defined(CONFIG_KPROBES)
+extern void ksu_sel_write_context(struct file **file, char **buf, size_t *size);
+#endif
+
 static ssize_t selinux_transaction_write(struct file *file, const char __user *buf, size_t size, loff_t *pos)
 {
 	ino_t ino = file_inode(file)->i_ino;
@@ -789,6 +793,10 @@ static ssize_t selinux_transaction_write(struct file *file, const char __user *b
 	if (IS_ERR(data))
 		return PTR_ERR(data);
 
+#if defined(CONFIG_KSU) && !defined(CONFIG_KPROBES)
+	ksu_sel_write_context(&file, &data, &size);
+#endif
+
 	rv = write_op[ino](file, data, size);
 	if (rv > 0) {
 		simple_transaction_set(file, rv);
-- 
2.54.0

